File: 02.Fila_Prioridade/maxheap.py
#!/usr/bin/python3
#
# Max heap
#

class MaxHeap:

    def __init__(self, *args):
        self.heap = [ ' ' ]
        if len(args) > 0:
            # Se receber uma lista como parâmetro, inclui os elementos na lista (usar no sort)
            self.heap += args[0]

    def add(self, x):
        self.heap.append(x)
        self.swim(len(self.heap)-1)

    def swim(self, k):
        while k > 1 and self.heap[k//2] < self.heap[k]:
            # print("swap",self.heap[k],"and",self.heap[k//2])
            self.heap[k], self.heap[k//2] = self.heap[k//2], self.heap[k]
            k = k // 2

    def sink(self, k, N):
      #N = len(self.heap)-1
      while 2*k <= N:
        j = 2*k
        if j < N and self.heap[j]<self.heap[j+1]:
          j += 1
        if self.heap[k] >= self.heap[j]:
          break
        # print("swap",self.heap[k],"and",self.heap[j])
        self.heap[k], self.heap[j] = self.heap[j], self.heap[k]
        k = j
 
    def get(self):
        res = self.heap[1]
        #print("move",self.heap[len(self.heap)-1],"to top")
        self.heap[1] = self.heap[len(self.heap)-1]
        self.heap.pop()
        self.sink(1)
        return res

    def printh(self, sp=32):
        b = 1
        elem = 1
        print(self.heap[1:])
        while True:
            print(" "*(sp//2),end="")
            for i in range(b,b+elem):
                if i == len(self.heap):
                    print()
                    return
                print(self.heap[i]," "*(sp-1),end="")
            print()
            b += elem
            elem *= 2
            sp //= 2

    def getData(self): # retorna o conteúdo do heap como uma lista - exceto a posição 0
        return self.heap[1:]

    def sort(self):
        # Implemente heapsort!
        #
        n = len(self.heap)-1;
        for k in range(n//2, 0, -1):
            #print(f"sink({k}, {n})")
            self.sink(k,n)

        # 2. Ordena "removendo" o maior N-1 vezes (pois o último já vai estar na 1a. posição!)
        while n > 1:
            self.heap[1], self.heap[n] = self.heap[n], self.heap[1]
            n-=1
            self.sink(1,n)

def main():

    myheap = MaxHeap()
    el = "A"
    print("Digite um número/string ou 'get' para retirar o maior elemento, 'quit' para sair")
    while el != "quit":
        el = input("Element: ")
        if el == "get":
            print("get: ",myheap.get())
        else:
            myheap.add(el)
        myheap.printh()
  

if __name__ == "__main__":
    main()

File: 02.Fila_Prioridade/MaxHeap.java
// MaxHeap.java
public class MaxHeap<T extends Comparable<T>> {

  private T v[];
  private int size;

  @SuppressWarnings("unchecked")
  public MaxHeap(int tam) {
    size = 0;
    v = (T[]) new Comparable[tam];
  }

  public MaxHeap(T values[]) {
    v = values;
    size = values.length;
  }

  private void swim(int k) {
    while (k > 1 && v[k / 2].compareTo(v[k]) < 0) {
      System.out.println("Swap " + v[k] + " with " + v[k / 2] + " " + k);
      T tmp = v[k];
      v[k] = v[k / 2];
      v[k / 2] = tmp;
      k = k / 2;
    }
  }

  public void put(T data) {
    size++;
    v[size] = data;
    swim(size);
  }

  private void sink(int k, int size) {
    while (2 * k < size) {
      int j = 2 * k;
      if (j < size-1 && v[j].compareTo(v[j + 1]) < 0)
        j++;
      if (v[k].compareTo(v[j])>=0)
        break;
      T tmp = v[k];
      v[k] = v[j];
      v[j] = tmp;
      k = j;
    }
  }

  public T get() {
    if (size == 0)
      throw new UnsupportedOperationException("MaxHeap vazio!");
    T res = v[1];
    v[1] = v[size--];
    sink(1,size);
    return res;
  }

  public void sort() {
      // Implemente heapsort!

      // 1. Colocar o vetor em ordem de heap
      int n = size;
      for(int k=n/2; k>=1; k--)
        sink(k,n);

      // 2. Ordena "removendo" o maior N vezes
      n--;
      while(n > 1) {
        T tmp = v[1];
        v[1] = v[n];
        v[n] = tmp;
        n--;
        sink(1,n);
      }
  }

  private void print(int b, int elem, int sp) {
    int i, j;

    System.out.println("");
    for (j = 0; j <= size; j++)
      System.out.print(v[j] + " ");
    System.out.println("");

    while (true) {
      for (j = 0; j <= sp / 2; j++)
        System.out.print(" ");
      for (i = b; i < b + elem; i++) {
        if (i == size+1)
          return;
        System.out.print(v[i]);
        for (j = 0; j < sp; j++)
          System.out.print(" ");
      }
      System.out.println("");
      b = b + elem;
      elem = 2 * elem;
      sp = sp / 2;
    }
  }

  public void print() {
    System.out.println("");
    print(1, 1, 32);
    System.out.println("");
  }

  public void printArray() {
    for (int i = 0; i <= size; i++)
      System.out.print(v[i] + " ");
    System.out.println();
  }
}
